docs: ✏️ fleek network how to deal with files using content addressable archives (#91)
FleekHQ/fleek-blog: content/posts/fleek-network-packing-content-addressed-data.md

When you use Fleek Network, you either provide your data packed into a format called a Content Archive ([CAR](https://ipld.io/specs/transport/car)) or an existing [CID](https://docs.ipfs.tech/concepts/content-addressing/#what-is-a-cid) of a CAR file, which hash addresses are unique and universally addressable 🛸.

## Immutability

Some of the principles that help us provide guarantees to end-users,
require a high-ability for content verification, as a consequence, immutability of files are critical to the system. To emphasize, immutability means the state of not changing, or being unable to change!

The web is nothing more and nothing less than a mirror of what it has become from within and among its main actors, the service providers, the central authorities 👮🏻 and the powers they have when managing user files.

Fleek Network deals with files in a manner where the content determines the address in which the user of the system can locate and verify it unquestionably 👩🏽‍⚖️. This is possible due to cryptography, of which the same data always produce the same hash deterministically.

> A file which content determines the hash, but also impossible to invert it. We shouldn't be able to reconstruct the data from a hash. It's unique, not two files produce the same file or content. Thus, a small change in the content should always generate a completely different hash.

Reviewer: ```suggestion
A file whose content determines the hash, but also impossible to invert it. We shouldn't be able to reconstruct the data from a hash. It's unique, not two files produce the same file or content. Thus, a small change in the content should always generate a completely different hash.
```

> In retrospect 👵🏼, what we have in the web today are files acessible via a URL address and the problem with this approach is that the content is not intrinsically tight to the address e.g. the content can change and the URL remaining the same. That is the problematic way we access files in the web today, that we call "Location addressing", and the way we solve it for the web of tomorrow, is called "Content addressing".

Reviewer: ```suggestion
In retrospect 👵🏼, what we have on the web today are files accessible via a URL address and the problem with this approach is that the content is not intrinsically tight to the address e.g. the content can change and the URL remains the same. That is the problematic way we access files on the web today, which we call "Location addressing", and the way we solve it for the web of tomorrow, is called "Content addressing".
```

> When content is immutable, we can verify its integrity and thus provide the ability get the content from anyone and everywhere. The ability to get it from anywhere, lead us to the decentralised and distributed nature of content storage and delivery in the Fleek Network.

## Content Addressing

> Content addressing is where we use a hash to access the content, and which allow us to verify that the content we received is the content we asked for 📢! For this we use a special hash called CID ([Content Identifier](https://docs.ipfs.tech/concepts/content-addressing/#what-is-a-cid)), a cryptography hash function that maps input of arbitrary size to output of a fixed size - the content identifiers are short, regardless of the size of the content, and the address does not actually tell us where the content is stored. It's also interesting to observe, that the CID is a sort of string like binary that is human-friendlier in comparison to the underlying binary, that is way longer 🤖.

Reviewer: ```suggestion
Content addressing is where we use a hash to access the content, and which allow us to verify that the content we received is the content we asked for 📢! For this we use a special hash called CID ([Content Identifier](https://docs.ipfs.tech/concepts/content-addressing/#what-is-a-cid)), a cryptography hash function that maps input of arbitrary size to output of a fixed size - the content identifiers are short, regardless of the size of the content, and the address does not actually tell us where the content is stored. It's also interesting to observe, that the CID is a sort of string like binary that is human-friendlier in comparison to the underlying binary, which is way longer 🤖.
```

> Caching and deduplication are possible due to immutability of content e.g. if content changes, let's say that an image has some new detail, the files share many of the same bytes. The amount of data we have to transfer to fetch is minimum, we'd only pull the difference. In today's web, we'd have to transfer both files in full, which is a worse path on resource allocation and performance.

## Hash functions

> The hash function for creating [CID's](https://docs.ipfs.tech/concepts/content-addressing/#what-is-a-cid) use `sha-256`, but there is support for other hashing algorithms, such as sha1 (used by Git), sha2-256, sha3-255, blake2b-160, etc. Some older algorithms are proven to not be collision free, so if algorithms can break, we have to swith the hash algorithm we use in the future! The problem with this switching of algorithms is the need of finding a future proof way of identifying the hash functions used to generate the hash, as well as the hash name.

Reviewer: ```suggestion
The hash function for creating [CID's](https://docs.ipfs.tech/concepts/content-addressing/#what-is-a-cid) use `sha-256`, but there is support for other hashing algorithms, such as sha1 (used by Git), sha2-256, sha3-255, blake2b-160, etc. Some older algorithms are proven not to be collision free, so if algorithms can break, we have to switch the hash algorithm we use in the future! The problem with this switching of algorithms is the need to find a future proof way of identifying the hash functions used to generate the hash, as well as the hash name.
```

> [Multihash](https://multiformats.io/multihash/) is a protocol and comes in to play to provide us the metadata useful for future proofing. 🙋‍♀️ To explain it in simple terms we'll provide an example, it is the composition where a hash is placed at the end, a prefix as a number to identify the algorithm used and a number to idenfify the hash name. Thereafter, we'd start raising some questions. With the simple example provided here, how would we get the data back without the ability to identity how it was encoded? Some users could use [cbor](https://en.wikipedia.org/wiki/CBOR), [protocol buffers](https://en.wikipedia.org/wiki/Protocol_Buffers), [json](https://en.wikipedia.org/wiki/JSON), etc; and there might be plenty of good reasons why of those choices. Maybe it's a compact binary encoding that is very efficient for storage, or easy to work with, etc.

Reviewer: ```suggestion
[Multihash](https://multiformats.io/multihash/) is a protocol that comes into play to provide us the valuable metadata for future proofing. 🙋‍♀️ To explain it in simple terms we'll provide an example, it is the composition where a hash is placed at the end, a prefix as a number to identify the algorithm used and a number to identify the hash name. Thereafter, we'd start raising some questions. With the simple example provided here, how would we get the data back without the ability to identify how it was encoded? Some users could use [cbor](https://en.wikipedia.org/wiki/CBOR), [protocol buffers](https://en.wikipedia.org/wiki/Protocol_Buffers), [json](https://en.wikipedia.org/wiki/JSON), etc; and there might be plenty of good reasons why of those choices. Maybe it's a compact binary encoding that is very efficient for storage, or easy to work with, etc.
```

> What's important is that it is the users choice and why [IPLD](https://ipld.io) becomes useful for Fleek Network's use-cases. A system for understanding and working with data, made up of a [Data Model](https://ipld.io/docs/intro/hello-world/#data-model-and-codecs) and [Codecs](https://ipld.io/docs/intro/hello-world/#data-model-and-codecs), some tools for [Linking](https://ipld.io/docs/intro/hello-world/#linking), and then a handful of other [Powerful Features](https://ipld.io/docs/intro/hello-world/#powerful-features) that help ups 👷 develop a decentralized application.

Reviewer: ```suggestion
What's important is that it is the user's choice and why [IPLD](https://ipld.io) becomes useful for Fleek Network's use-cases. A system for understanding and working with data, made up of a [Data Model](https://ipld.io/docs/intro/hello-world/#data-model-and-codecs) and [Codecs](https://ipld.io/docs/intro/hello-world/#data-model-and-codecs), some tools for [Linking](https://ipld.io/docs/intro/hello-world/#linking), and then a handful of other [Powerful Features](https://ipld.io/docs/intro/hello-world/#powerful-features) that help ups 👷 develop a decentralized application.
```

## Interplanetary linked data (IPLD)

> [Interplanetary linked data (IPLD)](https://ipld.io/) provide us all the metadata prefixes to soothe the system needs, provide us the data model of the content-addressable web, as discussed earlier. IPLD is a set of conventions for creating decentralised data-structures that are universally addressable and linkable.

Reviewer: ```suggestion
[Interplanetary linked data (IPLD)](https://ipld.io/) provides us all the metadata prefixes to soothe the system needs, and provides us the data model of the content-addressable web, as discussed earlier. IPLD is a set of conventions for creating decentralised data-structures that are universally addressable and linkable.
```

> These addressable and linkable data structures will allow us to do for data what URLs and links did for HTML web pages (Quote from [IPLD](https://ipld.io/docs/)).

## Content Addressable aRchive (CAR)

> For all the reasons demonstrated here 💁‍♀️, Fleek Network uses the IPLD CAR [Content Adressable aRchive](https://ipld.io/specs/transport/) to transport IPLD data. IPFS IPLD defines transport as of both file and stream format, meaning packing IPLD data together and interactivity that involves requests and responses.

Reviewer: ```suggestion
For all the reasons demonstrated here 💁‍♀️, Fleek Network uses the IPLD CAR [Content Addressable aRchive](https://ipld.io/specs/transport/) to transport IPLD data. IPLD defines transport as file and stream format, meaning packing IPLD data together and interactivity that involves requests and responses.
```

> As discussed above in [Hash functions -> multihash](#hash-functions), the CAR files contain data encoded in a particular codec, in the Fleek Network, we support any IPLD codec e.g. [dag-pb](https://ipld.io/docs/codecs/known/dag-pb/), which uses a stricter subset of Protocol Buffers to encode an object graph.

Reviewer: ```suggestion
As discussed above in [Hash functions -> multihash](#hash-functions), the CAR files contain data encoded in a particular codec, in the Fleek Network, we support any IPLD codec, e.g. [dag-pb](https://ipld.io/docs/codecs/known/dag-pb/), which uses a stricter subset of Protocol Buffers to encode an object graph.
```

To initialise IPFS, run:

```sh
ipfs init
```

Here's how our output looks like, yours will be slightly different but similar.

```sh
generating ED25519 keypair...done
peer identity: XXXXXXXXXXXXXXXXXXXXXXX
initializing IPFS node at /Users/<YOUR-USERNAME>/.ipfs
```

Find all subcommands available by running:

```
ipfs --help
```

Let's create a new file to use as an example afterwards, a file called `hello.txt` that has some content:

```sh
echo 'Hello world!' > hello.txt
```

Add the file by using the `add` subcommand, as follows:

```sh
ipfs add hello.txt
```

We're not running the IPFS daemon, it'll just add the file locally.

The output should look like:

```sh
added QmXgBq2xJKMqVo8jZdziyudNmnbiwjbpAycy5RbfDBoJRM hello.txt
 13 B / 13 B [==================================================] 100.00%
```

The CID for our `hello.text` is `QmXgBq2xJKMqVo8jZdziyudNmnbiwjbpAycy5RbfDBoJRM`, this is the content's cryptographic hash. If the file content changes, the hash will change, otherwise if the file's content's the same, the hash will always be the same, as described [here](#immutability).

> 💡 Notice that our string starts with a "Qm" which refers to the v0 of CID. The v1 of CID starts with "Bafy", or "Bafk" sometimes. You can learn a lot from the CID by using the [cid.ipfs.tech](https://cid.ipfs.tech/#QmXgBq2xJKMqVo8jZdziyudNmnbiwjbpAycy5RbfDBoJRM) tool. If we wish, we can do [CID conversion](https://docs.ipfs.tech/concepts/content-addressing/#cid-conversion) from v0 to v1 using the ipfs cli.

Reviewer: ```suggestion
💡 Notice that our string starts with a "Qm" which refers to the v0 of CID. The v1 of CID starts with "Bafy", or "Bafk" sometimes. You can learn a lot from the CID by using the [cid.ipfs.tech](https://cid.ipfs.tech/#QmXgBq2xJKMqVo8jZdziyudNmnbiwjbpAycy5RbfDBoJRM) tool. We can do [CID conversion](https://docs.ipfs.tech/concepts/content-addressing/#cid-conversion) from v0 to v1 using the ipfs cli.
```
